lib/puppet/indirector/rest.rb

```diff
@@ -277,6 +281,29 @@ def parse_response(response)
     end
   end
 
+  # This is a copy of convert_to_http_error above, but adapted to be
+  # used with Puppet::HTTP::ResponseErrors
+  def handle_response_error(url, request, response)
+    if is_http_404?(response)
+      return nil unless request.options[:fail_on_404]
+
+      _, body = parse_response(response)
+      msg = _("Find %{uri} resulted in 404 with the message: %{body}") % { uri: elide(url.path.to_s, 100), body: body }
```

Comment:
`_("Error %{code} on SERVER: %{returned_message}") % { code: response.code, returned_message: returned_message }`

Though the original `find` method does error with this message, the save method errors with ^^, which is more generic. If we need to maintain consistency in these error messages, it's probably best to go with this one.

Reply:
on second pass though, `save` returns `nil` on 404, regardless of what `request.options[:fail_on_404]` is, which is fun. This probs needs another pass :o(

Reply:
Each rest terminus does the right thing now, though I should be able to DRY some of that.